### ISSUE TITLE
Problem with Terminal and linux shell different from bash

### ISSUE THREAD

**Issue by odedonato** [argoproj/argo-cd#9641]
**Describe the bug**

Tried the new most wanted feature, Terminal, it works with linux container with `bash` but got some problem with container `Alpine` based.

In screenshot, tried on `busybox:stable` container (it's in loop to keep it running) got # after `OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "bash": executable file not found in $PATH: unknown` but cannot use it (cannot write)

Tried with some other container based on `node:lts-alpine` and got only black screen with blinking cursor.

Shell is working with same container using `Rancher` and `Lens`.

**To Reproduce**

Click on Terminal on Alpine based container

**Expected behavior**

Get shell from container and possibility to use it

**Screenshots**
<img width="1431" alt="image" src="https://user-images.githubusercontent.com/3715824/173307885-64827f86-30d0-4d06-8547-368215cdc8be.png">

**Version**

```shell
2.4.0+91aefab
```

**Comment by mvpwar** [comment]
use nginx image instead, it works

**Comment by mihma** [comment]
Same issue for me when using `node:14.16.0-alpine` base image

**Comment by crenshaw-dev** [comment]
Can y'all check the browser's console and/or network dev tools for errors when connecting to the alpine pod?

**Comment by mihma** [comment]
Hi @crenshaw-dev ,
I get blinking cursor with never dropping connection to the terminal as well as some weird font download error.

PS. It works fine on non-alpine pods: working terminal and zero font errors.

<img width="1242" alt="Screenshot 2022-06-14 at 15 45 11" src="https://user-images.githubusercontent.com/356633/173593697-eaf60050-1a63-470c-8b58-ff1464c34354.png">

**Comment by arthurk** [comment]
It fails for our alpine based images too. But I noticed that the web terminal works when bash is installed via `apk add bash`

**Comment by djfinnoy** [comment]
Same issue here, works fine with containers that have bash, weird websocket issues with containers that only have sh. Black screen for containers with neither.

**Comment by crenshaw-dev** [comment]
From #9425 :

> Looks like ArgoCD is having issues spawning a terminal with the alpine ash shell.
> 
> It works after adding bash package:
> 
> apk add bash

Is it possible we just need to add `ash` as a shell to try here?

https://github.com/argoproj/argo-cd/blob/56eb09507e917de9c962a3249435d05f4dbf7e4b/server/application/terminal.go#L219

**Comment by odedonato** [comment]
@crenshaw-dev I was wondering if is useful and possible to put validShells in a ConfigMap, so we can add our shell, like pwsh, ash, etc.
In that way is not binded in the code.

**Comment by crenshaw-dev** [comment]
@odedonato making it configurable is probably a good idea. But I'm not yet convinced that the (only) problem is not having enough shells in that array. If someone has time, a quick test to see if adding `ash` fixes Alpine would be helpful.

**Comment by odedonato** [comment]
@crenshaw-dev I downloaded the code and added `ash` then builded the container and tried but didn't work, same black screen with blinking cursor.

**Comment by arthurk** [comment]
I don't think adding more shells to the list is going to fix the problem. I often use k9s to launch shells in containers and it works  well. Their code looks like this:

https://github.com/derailed/k9s/blob/fc8ffe5d37d068f14e86aed7e8b847d016c99b19/internal/view/exec.go#L30

I would be interested to see if putting `sh` before `bash` would show a different result.

**Comment by odedonato** [comment]
@crenshaw-dev turns out that the black screen it's only on local k8s cluster where is located argocd and it's because it was a lack of permission, added in `ClusterRole:argocd-server`

```- apiGroups:
  - ""
  resources:
  - pods/exec
  verbs:
  - create```

  After that I got the error `OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "bash": executable file not found in $PATH: unknown` same error in the screenshot in the first post, got # but not usable, so I tried what suggested by @arthurk.

  I built argo with `sh` first and it works, but it opens only `sh` on all containers. It's just a workaroud imho.

  I found this: https://github.com/argoproj/argo-cd/blob/56eb09507e917de9c962a3249435d05f4dbf7e4b/server/application/terminal.go#L225

**Comment by yeya24** [comment]
I am aware of this issue but there might be other isssue more than the shell order. For k8s dashboard code, it uses the same shell order but it can enable exec for alpine images using sh.

Wondering if this is caused by error check at the frontend side.

**Comment by elmariofredo** [comment]
It feels like this issue will be bit harder to figure out, as there are multiple libraries involved. My suggestion is to quick fix this issue, by temporary removing bash shell from list which would resolve this issue for most containers. When problem shell test is resolved we can put bash back. WDYT?

**Comment by yeya24** [comment]
> It feels like this issue will be bit harder to figure out, as there are multiple libraries involved. My suggestion is to quick fix this issue, by temporary removing bash shell from list which would resolve this issue for most containers. When problem shell test is resolved we can put bash back. WDYT?

I don't have any preference. Feel free to open a pr to change the order and ArgoCD maintainers can decide whether they are willing to do this change or not. Again maybe a configmap for shells so that users can reorder the shells as they want would be more flexible.

**Comment by crenshaw-dev** [comment]
I’m in favor of that change, and I agree that a configurable list makes sense.

**Comment by m-yosefpor** [comment]
all the following 3 issues are the same:
- https://github.com/argoproj/argo-cd/issues/9643
- https://github.com/argoproj/argo-cd/issues/9641
- https://github.com/argoproj/argo-cd/issues/9605


I think we can follow in one thread.

**Comment by m-yosefpor** [comment]
> I’m in favor of that change, and I agree that a configurable list makes sense. I think the config can go in ‘argocd-cmd-params-cm’.

I'm not against this feature, but I feel  this feature might not be useful for most of the use-cases, as logical order of shells for a simple pod terminal is what currently listed in the code (as `bash` and `sh` are the most common shells which could be found in container images). Also it doesn't really fixes the retrying next shell issue (which is the actual problem here), since as I wrote in [here](https://github.com/argoproj/argo-cd/issues/9605#issuecomment-1160140844), we might still want `bash` to be the first container and retry `sh` in case it is missing. Having a global config for it doesn't really solve this exact problem (of course it could be a tmp workaround).

**Comment by crenshaw-dev** [comment]
> of course it could be a tmp workaround

I'm satisfied with a temporary workaround for now. This will solve a headache for 90% of use cases and will buy us time to figure out the 10%.

Making the list configurable is a win regardless, because people might want to add options that are not currently in the hardcoded list (e.g. `ash`).

**Comment by elmariofredo** [comment]
> I think I actually found the real issue. Can someone test the PR to confirm? #9895

I can confirm that your PR is solving the issue. Even that bash will fail, switch to sh is working and terminal is usable 🎉 

Thanks!

<img width="1071" alt="image" src="https://user-images.githubusercontent.com/82104/177723328-f5860b92-1d43-4c85-847e-94c047db1d30.png">

**Comment by arthurk** [comment]
@michalziobro the fix is not in 2.4.4, I think it will be available in the next release

**Comment by crenshaw-dev** [comment]
Yep, I'll release the fix in the next few hours. It'll be in 2.4.6, since I've got some other stuff queued up for 2.4.5.

**Comment by arthurk** [comment]
I've upgraded to 2.4.6 and can confirm that the web terminal feature works well now

**Comment by S0n98** [comment]
I'm also facing an issue with web shell, got only black screen with blinking cursor, but the logs I got from argocd-server are quite different:

```
2022/12/21 08:03:07 http: response.WriteHeader on hijacked connection from github.com/argoproj/argo-cd/v2/server/application.(*terminalHandler).ServeHTTP (terminal.go:237)
71
2022/12/21 08:03:07 http: response.Write on hijacked connection from fmt.Fprintln (print.go:265)
```
image used: `linuxserver/openvpn-as:latest`
